README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# GoodFitSBM: Monte Carlo goodness-of-fit tests for Stochastic Blockmodels (SBMs)

<!-- <img src="man/figures/logo.png" align="right" width="150"/> -->
<!-- [![CRAN_Status_Badge](https://img.shields.io/cran/v/MatchIt?color=952100)](https://cran.r-project.org/package=lmw) [![CRAN_Downloads_Badge](https://cranlogs.r-pkg.org/badges/MatchIt?color=952100)](https://cran.r-project.org/package=lmw) -->

### NEWS

GoodFitSBM (Version 0.0.1): `GoodFitSBM` comprises functionality that
performs the *goodness-of-fit* test for a **beta-SBM** (one of the three
variants of SBMs as discussed in [Karwa et
al. (2023)](https://doi.org/10.1093/jrsssb/qkad084), used for modelling
network data).

### Note

The math rendering of `README.md` is not ideal, as a result of which
most of the math notations and equations remains un-rendered, hence
`knit` the `README.Rmd`, or refer to the PDF attached as
[README.pdf](https://drive.google.com/file/d/1-mzMOdQ5zWv4uIfoegl1kTGAUPPdd00j/view?usp=sharing).

### Overview

*Stochastic blockmodels* (SBMs) contributed to the theoretical and
algorithmic developments for analyzing *network* data, which has been
facilitated by the availability of data in diverse fields like *social
sciences*, *web recommender systems*, *protein networks*, *genomics*,
and *neuroscience*. SBMs being the generalization of the Erdős-Rényi
model given by [Erdős and Rényi
(1960)](https://doi.org/10.1515/9781400841356.38) was proposed
originally in context of *social sciences* for directed and undirected
graphs, whereas now it has been vastly extended and utilized in *latent
blocks in undirected graphs*, *latent space models*, *variable degree
distribution*, *dynamically evolving networks*, etc., becoming one of
the more popular approaches to model network data in computer science,
statistics and machine learning.

[Karwa et al. (2023)](https://doi.org/10.1093/jrsssb/qkad084) addresses
a very important aspect of *model fitting* by constructing (exact)
goodness-of-fit tests (under finite-sample setting) for three variants
of SBMs respectively used for modeling network data (where model
adequacy procedures are somewhat elusive in general), viz., *Erdős-Rényi
SBM* (ER-SBM), *Additive SBM*, and *beta-SBM*, where the main idea
revolves around a *frequentist conditional goodness-of-fit test*
conditioned on a sufficient statistic ([Karwa et
al. (2023)](https://doi.org/10.1093/jrsssb/qkad084) also illustrates its
Bayesian counterpart).

### Intended Use of the package `GoodFitSBM`

Out of the three variants of the SBMs in ([Karwa et
al. (2023)](https://doi.org/10.1093/jrsssb/qkad084)) to model network
data, `GoodfitSBM` addresses goodness-of-fit test under the framework of
a **beta-SBM**. With a focus on *simple undirected*, and *unweighted*
networks (graphs) having *no self-loop*, the package comprises of four
functions viz., `get_mle()`, `goftest()`, `graphchi()`, and
`sample_a_move()` - among which `goftest()` performs the major
functionality of performing the goodness-of-fit test, in process,
obtains the value of the chi-square test statistic and the corresponding
$p-$value (using a Monte Carlo approach), after proper sampling of the
graph (a Markov move or basis) - done by the function `sample_a_move()`
under the beta-SBM framework. Computation of the chi-square test
statistic value under the beta-SBM framework is done by `graphchi()`,
which in turn requires the estimates (maximum likelihood estimation
(MLE) in our case) of the edge probabilities $q_{ij}$ between block
$B_i$ and block $B_j$ as, $\widehat{q}_{ij}$, which is done by the
function `get_mle()`.

The corresponding $p-$value obtained from `goftest()` are further
analysed to examine the extent of fit of the beta-SBM to the given
network (graph) (usual rule applies to reject the null of a good fit
when, $p \leq \alpha$ at level $0< \alpha < 1$). There are other
functions collated, each performing the required functionality in the
process, see the `R` files in [GoodFitSBM Github
Repo](https://github.com/Roy-SR-007/GoodFitSBM/tree/master/R) for
details.

### The beta-SBM and Related Goodness-of-Fit (GoF) Framework

In this section, we outline the theoretical part on which the entire
package `GoodFitSBM` is based.

#### Stochastic Blockmodels (SBMs)

Consider $G$ to be a graph on $n$ nodes, $g$ being its realization. It
is assumed that, all graphs are *unweighted* and *undirected*, and there
are *no self-loops*. Therefore, the graph $G$ can also be referred to by
its $n\times n$ adjacency matrix, $\mathbf{g}$, where $g_{uv} = 1$, if
there is an edge between node $u$ and $v$ in the graph $\mathbf{g}$, and
$0$ otherwise.

An *exponential family random graph model (ERGM)* assumes that the
probability of observing a graph $\mathbf{g}$ depends only on a vector
of sufficient statistics, $T(\mathbf{g})$, i.e., the probability of
observing a given network $G=\mathbf{g}$ is,

$$
\mathbb{P}_{\theta}(G=\mathbf{g}) = \frac{exp{\langle T(\mathbf{g}), \theta\rangle}}{\psi(\theta)}
$$ where
$\psi(\theta) = \sum_{\mathbf{g}}exp(\langle T(\mathbf{g}), \theta\rangle)$
is the normalizing constant, $\theta \in \Theta$ is a vector of natural
parameters, and $T(\mathbf{g})$ is the vector of minimal sufficient
statistics for the underlying model. Let $p_{uv}$ be the probability of
an edge between node $u$ and node $v$, where it is assumed that,
$g_{uv}\sim \mbox{Bernoulli}(p_{uv})$.

A *stochastic blockmodel (SBM)* postulates that the nodes are
partitioned into $k$ blocks and the probability $p_{uv}$ depends on
their block membership, where ([Karwa et
al. (2023)](https://doi.org/10.1093/jrsssb/qkad084)) considers three
different *log-linear* parametrizations of $p_{uv}$, one of which yields
the beta-SBM.

#### The beta-SBM

Beta-SBM is exponential family version of the degree-corrected
stochastic blockmodels suggested by [Karrer and Newman
(2011)](https://link.aps.org/doi/10.1103/PhysRevE.83.016107). The
log-linear model for the edge probabilities is parametrized by
$\binom{k+1}{2}$ block parameters $\alpha_{z(u)z(v)}$ and $n$
node-specific parameters $\beta_{u}$ for $1\leq u \leq n$, with the
log-odds of the probability of an edge $\{u,v\}$ given by,

$$
\log\bigg(\frac{p_{uv}}{1 - p_{uv}}\bigg) = \alpha_{z(u)z(v)} + \beta_{u} + \beta_{v}
$$ When $\mathcal{Z}$ (the block assignments) are known, the beta-SBM is
an exponential family model with natural parameter vector
$\theta = (\beta, \alpha)$, where $\beta = (\beta_1, \ldots, \beta_{n})$
and $\alpha$ is the vector of the upper diagonal elements of the
$k\times k$ matrix $((\alpha_{i,j}))$. The natural parameter space is
$\Theta \equiv \mathbb{R}^n \times \mathbb{R}^{\binom{k+1}{2}}$. The
vector of sufficient statistics $T_{\beta}$ contains the degree of each
node $i\in [n]$ and the number of edges between each pair of blocks
$B_{i}$ and $B_{j}$ for $1\leq i < j \leq k$.

#### Goodness-of-Fit Test for the beta-SBM

Let us define the goodness-of-fit test as,

$$
H_0: G \sim \mathbb{P}_{\theta}(G|Z)
$$ against general alternatives, where $\mathbb{P}_{\theta}(G|Z)$ is a
variant of the SBM with a generic parameter vector $\theta$ and block
assignment $Z$ (it is assumed that the number of blocks $k$ is fixed and
known).

Considering $T(\mathbf{g})$ to be the vector of sufficient statistics in
$\mathbb{P}_{\theta}(G|Z)$, define a subset of the sample space as,
$F_{u} := \{\mathbf{g}:T(\mathbf{g}) = u\}$, which is the known as the
*fiber* of $u$ under the given exponential family model. Now the usual
chi-square statistic (for the goodness-of-fit test) is not constant on
the above defined fibers of the beta-SBM variant. Thus, analogous to the
beta-model in [Petrovic et al. (2010)](https://arxiv.org/abs/0909.0073),
we use the Pearson’s chi-square statistic,

$$
GoF_{Z}(\mathbf{g}) = \chi^{2}_{\mbox{beta-SBM}}(\mathbf{g}, Z) = \sum_{1\leq u < v \leq n}\frac{(g_{uv} - \widehat{g}_{uv})^2}{\widehat{g}_{uv}}
$$ where,
$\widehat{g}_{uv} = exp(\widehat{\alpha}_{z(u)z(v)} + \widehat{\beta}_{u} + \widehat{\beta}_{v})/( exp(\widehat{\alpha}_{z(u)z(v)} + \widehat{\beta}_{u} + \widehat{\beta}_{v}))$,
where the MLEs $\widehat{\alpha}$ and $\widehat{\beta}$ are associated
with the MLEs of the edge probabilities $q_{uv}$, as obtained in the
`get_mle()` function. Large values of $\chi^{2}_{\mbox{beta-SBM}}$
correspond to departure from the null $(H_0)$ as stated above.

- For the algorithm of sampling of the graph (as done by
  `sample_a_move()`) - or a Markov move (basis) - and other details, see
  Section 5.3 of ([Karwa et
  al. (2023)](https://doi.org/10.1093/jrsssb/qkad084)).

### Use and Related Functionalities

In this section, we highlight the installation and different
functionalities included in `GoodFitSBM` along with its implementation
on some test cases and a real-life dataset.

#### Installation

To install and load-up the (development version 0.0.1) package
`GoodFitSBM` from [GoodFitSBM Github
Repo](https://github.com/Roy-SR-007/GoodFitSBM), run the following
commands.

``` r
# install.packages("devtools")
# install.packages("remotes")
remotes::install_github("Roy-SR-007/GoodFitSBM")
```

    #> 
    #> ── R CMD build ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
    #> * checking for file ‘/private/var/folders/ls/s91_zt990n9b6gdbr57m19900000gn/T/RtmpOAPRU6/remotes35cb9537541/Roy-SR-007-GoodFitSBM-f35c19f/DESCRIPTION’ ... OK
    #> * preparing ‘GoodFitSBM’:
    #> * checking DESCRIPTION meta-information ... OK
    #> * checking for LF line-endings in source and make files and shell scripts
    #> * checking for empty or unneeded directories
    #> * building ‘GoodFitSBM_0.0.1.tar.gz’

``` r
library(GoodFitSBM)
```

#### Example 1: Sampling a Graph using `sample_a_move()`

Here we consider sampling (Markov move) a graph (under the beta-SBM
framework) with a total of $n=150$ nodes, with $k = 3$ blocks of sizes
$50$ each.

``` r
library(igraph)
RNGkind(sample.kind = "Rounding")
set.seed(1729)

# We model a network with 3 even classes
n1 = 50
n2 = 50
n3 = 50

# Generating block assignments for each of the nodes
n = n1 + n2 + n3
class = rep(c(1, 2, 3), c(n1, n2, n3))

# Generating the adjacency matrix of the network
# Generate the matrix of connection probabilities
cmat = matrix(
  c(
    30, 0.05, 0.05,
    0.05, 30, 0.05,
    0.05, 0.05, 30
  ),
  ncol = 3,
  byrow = TRUE
)
pmat = cmat / n

# Creating the n x n adjacency matrix
adj <- matrix(0, n, n)
for (i in 2:n) {
  for (j in 1:(i - 1)) {
    p = pmat[class[i], class[j]] # We find the probability of connection with the weights
    adj[i, j] = rbinom(1, 1, p) # We include the edge with probability p
  }
}

adjsymm = adj + t(adj)

# graph from the adjacency matrix
G = igraph::graph_from_adjacency_matrix(adjsymm, mode = "undirected", weighted = NULL)

# plotting the current graph
plot(G, main = "The current graph")
```

<img src="man/figures/README-2-1.png" style="display: block; margin: auto;" />

``` r
# sampling a Markov move for the beta SBM
G_sample = sample_a_move(class, G)

# plotting the sampled graph
plot(G_sample, main = "The sampled graph after one Markov move for beta SBM")
```

<img src="man/figures/README-2-2.png" style="display: block; margin: auto;" />

#### Example 2: Estimating Edge Probabilities using `get_mle()`

Here we determine the MLEs of the edge probabilities
$(\widehat{q}_{uv})$ for a graph with $k=3$ blocks each of size $2$, and
having a total of $n=6$ nodes.

``` r
library(igraph)
RNGkind(sample.kind = "Rounding")
set.seed(1729)

# We model a network with 3 even classes
n1 = 2
n2 = 2
n3 = 2

# Generating block assignments for each of the nodes
n = n1 + n2 + n3
class = rep(c(1, 2, 3), c(n1, n2, n3))

# Generating the adjacency matrix of the network
# Generate the matrix of connection probabilities
cmat = matrix(
  c(
    30, 0.50, 0.50,
    0.50, 30, 0.50,
    0.50, 0.50, 30
  ),
  ncol = 3,
  byrow = TRUE
)
pmat = cmat / n

# Creating the n x n adjacency matrix
adj <- matrix(0, n, n)
for (i in 2:n) {
  for (j in 1:(i - 1)) {
    p = pmat[class[i], class[j]] # We find the probability of connection with the weights
    adj[i, j] = rbinom(1, 1, p) # We include the edge with probability p
  }
}

adjsymm = adj + t(adj)

# graph from the adjacency matrix
G = igraph::graph_from_adjacency_matrix(adjsymm, mode = "undirected", weighted = NULL)

# mle of the edge probabilities
get_mle(G, class)
```

    #> 2 iterations: deviation 5.551115e-16

    #>            [,1]       [,2]       [,3]       [,4]       [,5]       [,6]
    #> [1,] 0.08333333 0.08333333 0.08333333 0.08333333 0.08333333 0.08333333
    #> [2,] 0.08333333 0.08333333 0.08333333 0.08333333 0.08333333 0.08333333
    #> [3,] 0.08333333 0.08333333 0.00000000 0.00000000 0.08333333 0.08333333
    #> [4,] 0.08333333 0.08333333 0.00000000 0.00000000 0.08333333 0.08333333
    #> [5,] 0.08333333 0.08333333 0.08333333 0.08333333 0.08333333 0.08333333
    #> [6,] 0.08333333 0.08333333 0.08333333 0.08333333 0.08333333 0.08333333

#### Example 3: Computing the chi-square test statistic value using `graphchi()`

Computing the chi-square test statistic value for a network with a total
of $n=9$ nodes, and $k=3$ blocks of size $3$ each.

``` r
library(igraph)
RNGkind(sample.kind = "Rounding")
set.seed(1729)

# We model a network with 3 even classes
n1 = 3
n2 = 3
n3 = 3

# Generating block assignments for each of the nodes
n = n1 + n2 + n3
class = rep(c(1, 2, 3), c(n1, n2, n3))

# Generating the adjacency matrix of the network
# Generate the matrix of connection probabilities
cmat = matrix(
  c(
    30, 0.5, 0.5,
    0.5, 30, 0.5,
    0.5, 0.5, 30
  ),
  ncol = 3,
  byrow = TRUE
)
pmat = cmat / n

# Creating the n x n adjacency matrix
adj <- matrix(0, n, n)
for (i in 2:n) {
  for (j in 1:(i - 1)) {
    p = pmat[class[i], class[j]] # We find the probability of connection with the weights
    adj[i, j] = rbinom(1, 1, p) # We include the edge with probability p
  }
}

adjsymm = adj + t(adj)

# graph from the adjacency matrix
G = igraph::graph_from_adjacency_matrix(adjsymm, mode = "undirected", weighted = NULL)

# mle of the edge probabilities
p.hat = get_mle(G, class)
```

    #> 2 iterations: deviation 6.661338e-16

``` r
# chi-square test statistic values
graphchi(G, class, p.hat)
```

    #> [1] 16.66667

#### Main Example: `goftest()` in action

- **Instance 1**: Here we consider testing for the goodness-of-fit to
  the beta-SBM modeling a graph (network) with $k=3$ blocks of size $50$
  each, and a total of $n=150$ nodes.

``` r
library(igraph)
RNGkind(sample.kind = "Rounding")
set.seed(1729)

# We model a network with 3 even classes
n1 = 50
n2 = 50
n3 = 50

# Generating block assignments for each of the nodes
n = n1 + n2 + n3
class = rep(c(1, 2, 3), c(n1, n2, n3))

# Generating the adjacency matrix of the network
# Generate the matrix of connection probabilities
cmat = matrix(
  c(
    30, 0.05, 0.05,
    0.05, 30, 0.05,
    0.05, 0.05, 30
  ),
  ncol = 3,
  byrow = TRUE
)
pmat = cmat / n

# Creating the n x n adjacency matrix
adj <- matrix(0, n, n)
for (i in 2:n) {
  for (j in 1:(i - 1)) {
    p = pmat[class[i], class[j]] # We find the probability of connection with the weights
    adj[i, j] = rbinom(1, 1, p) # We include the edge with probability p
  }
}

adjsymm = adj + t(adj)

# When class assignment is known
out = goftest(adjsymm, C = class, numGraphs = 100)
```

    #> 2 iterations: deviation 1.580815e-10

``` r
chi_sq_seq = out$statistic
pvalue = out$p.value
print(pvalue)
```

    #> [1] 0.2

``` r
# Plotting histogram of the sequence of the test statistics
hist(chi_sq_seq, 20, xlab = "chi-square test statistics", main = NULL, ylim = c(0, 30))
# adding test statistic on the observed network
abline(v = chi_sq_seq[1], col = "red", lwd = 5)
legend("topleft", legend = paste("observed GoF = ", chi_sq_seq[1]))
```

<img src="man/figures/README-5-1.png" style="display: block; margin: auto;" />

From the $p-$value obtained viz., $p=0.20 >> \alpha = 0.05$, we fail to
the reject the null of a good fit of the given (observed) network to the
beta-SBM, at level $\alpha = 0.05$.

- **Instance 2**: Here we consider testing for the goodness-of-fit to
  the beta-SBM modeling a graph (network) with $k=3$ blocks with sizes
  $30$, $20$, and $50$, i.e., with a total of $n=100$ nodes.

``` r
library(igraph)
library(igraph)
RNGkind(sample.kind = "Rounding")
set.seed(1729)

# We model a network with 3 even classes
n1 = 30
n2 = 20
n3 = 50

# Generating block assignments for each of the nodes
n = n1 + n2 + n3
class = rep(c(1, 2, 3), c(n1, n2, n3))

# Generating the adjacency matrix of the network
# Generate the matrix of connection probabilities
cmat = matrix(
  c(
    30, 0.05, 0.05,
    0.05, 30, 0.05,
    0.05, 0.05, 30
  ),
  ncol = 3,
  byrow = TRUE
)
pmat = cmat / n

# Creating the n x n adjacency matrix
adj <- matrix(0, n, n)
for (i in 2:n) {
  for (j in 1:(i - 1)) {
    p = pmat[class[i], class[j]] # We find the probability of connection with the weights
    adj[i, j] = rbinom(1, 1, p) # We include the edge with probability p
  }
}

adjsymm = adj + t(adj)

# When class assignment is known
out = goftest(adjsymm, C = class, numGraphs = 100)
```

    #> 2 iterations: deviation 3.353762e-11

``` r
chi_sq_seq = out$statistic
pvalue = out$p.value
print(pvalue)
```

    #> [1] 0.73

``` r
# Plotting histogram of the sequence of the test statistics
hist(chi_sq_seq, 20, xlab = "chi-square test statistics", main = NULL, ylim = c(0, 30))
# adding test statistic on the observed network
abline(v = chi_sq_seq[1], col = "red", lwd = 5)
legend("topleft", legend = paste("observed GoF = ", chi_sq_seq[1]))
```

<img src="man/figures/README-6-1.png" style="display: block; margin: auto;" />

Observe that, the $p-$value obtained is much higher as compared to the
previous instance viz., $p=0.73 >>> \alpha = 0.05$, hence we fail to the
reject the null of a good fit of the given (observed) network to the
beta-SBM, at level $\alpha = 0.05$.

#### Application to Real Datasets: Zachary’s Karate Club Data

*Zachary’s Karate Club Data* is a classic, well-studied social network
of friendships between 34 members of a Karate club at a US university,
collected by Wayne Zachary in 1977 ([Zachary
(1977)](http://vlado.fmf.uni-lj.si/pub/networks/data/Ucinet/UciData.htm#zachary)).

Here, we consider fitting the beta-SBM model with $k = 2$ blocks, each
with sizes $10$ and $24$ respectively, i.e., a total of $n=34$ nodes.

``` r
library(igraph)

set.seed(334003213)

data("zachary")

d = zachary # the Zachary's Karate Club data set

# the adjacency matrix
A_zachary = as.matrix(d[1:34, ])
colnames(A_zachary) = 1:34

# obtaining the graph from the adjacency matrix above
g_zachary = igraph::graph_from_adjacency_matrix(A_zachary, mode = "undirected",
                                                weighted = NULL)

# plotting the graph (network) obtained
plot(g_zachary,
main = "Network (Graph) for the Zachary's Karate Club data set")
```

<img src="man/figures/README-7-1.png" style="display: block; margin: auto;" />

``` r
# block assignments
K = 2 # no. of blocks

n1 = 10
n2 = 24
n = n1 + n2

# known class assignments
class = rep(c(1, 2), c(n1, n2))
# goodness-of-fit tests for the Zachary's Karate Club data set
out_zachary = goftest(A_zachary, C = class, numGraphs = 100)
```

    #> 2 iterations: deviation 3.410605e-13

``` r
chi_sq_seq = out_zachary$statistic
pvalue = out_zachary$p.value
print(pvalue)
```

    #> [1] 0

``` r
# Plotting histogram of the sequence of the test statistics
hist(chi_sq_seq, 20, xlab = "chi-square test statistics", main = NULL, ylim = c(0, 30))
# adding test statistic on the observed network
abline(v = chi_sq_seq[1], col = "red", lwd = 5)
legend("topleft", legend = paste("observed GoF = ", chi_sq_seq[1]))
```

<img src="man/figures/README-7-2.png" style="display: block; margin: auto;" />

Note that, from the $p-$value obtained as well as the observed value of
the chi-square test statistic (plotted histogram), we reject the null of
a good fit, i.e., the data (network/graph) does not fit the beta-SBM at
all; quite similar to the lines of conclusion as claimed by ([Karwa et
al. (2023)](https://doi.org/10.1093/jrsssb/qkad084)) under the framework
of an ER-SBM, see Section 7.1.

#### An Addendum to Unknown Block Assignments

All of the above examples along with the instances worked-out with the
lines of code, deals with the situation where the *block assignments are
known*. On the contrary, if the block assignments are *unknown* (with
the number of blocks always fixed and known), we refer to ([Qin and Rohe
(2013)](https://proceedings.neurips.cc/paper_files/paper/2013/file/0ed9422357395a0d4879191c66f4faa2-Paper.pdf))
for estimating the block assignments through *regularized spectral
clustering*. See the following lines of code.

``` r
library(igraph)

set.seed(100)
# goodness-of-fit tests for the Zachary's Karate Club data set; 
# with unknown block assignments
out_zachary_unknown = goftest(A_zachary, K = 2, C = NULL, numGraphs = 100)
```

    #> 2 iterations: deviation 2.984279e-13

``` r

NEWS.md
# *News*

# GoodFitSBM (Version 1)

---

`GoodFitSBM` comprises functionality that performs the *goodness-of-fit* test for an **beta-SBM** (one of the three variants of SBMs discussed in [Karwa et al. (2023)](https://doi.org/10.1093/jrsssb/qkad084) used for modelling network data.


R/GOFTest_betaSBM.R

#'
#' @title Monte Carlo goodness-of-fit tests for Stochastic Blockmodels (SBMs)
#'
#' @description `goftest` performs chi square goodness-of-fit test for network data considering the model as beta SBM (Karwa et al. (2023))
#'
#' @param A n by n binary symmetric adjacency matrix representing an undirected graph where n is the number of nodes in the graph
#' @param K positive integer scalar representing the number of blocks; K>1
#' @param C positive integer vector of size n for block assignments of each node; from 1 to K (no of blocks)
#' @param numGraphs number of graphs to be sampled; default value is 100
#'
#' @return A list with the elements
#' \item{statistic}{the values of the chi-square test statistics on each sampled graph}
#' \item{p.value}{the p-value for the test}
#'
#' @importFrom igraph graph.empty
#' @importFrom igraph vcount
#' @importFrom igraph graph
#' @importFrom igraph ecount
#' @importFrom igraph graph.intersection
#' @importFrom igraph graph.difference
#' @importFrom igraph as.directed
#' @importFrom igraph is.simple
#' @importFrom igraph is.directed
#' @importFrom igraph graph.union
#' @importFrom igraph get.edges
#' @importFrom igraph get.edge.ids
#' @importFrom igraph as.undirected
#' @importFrom igraph get.edgelist
#' @importFrom igraph subgraph.edges
#'
#' @include Estimation_MLE.R
#' @include TestStatistic_graphchi.R
#' @include Sampling_Graph.R
#' @include Estimation_Block.R
#'
#' @export
#'
#' @examples
#'
#' # Example 1
#'
#' RNGkind(sample.kind = "Rounding")
#' set.seed(1729)
#'
#' # We model a network with 3 even classes
#' n1 = 50
#' n2 = 50
#' n3 = 50
#'
#' # Generating block assignments for each of the nodes
#' n = n1 + n2 + n3
#' class = rep(c(1, 2, 3), c(n1, n2, n3))
#'
#' # Generating the adjacency matrix of the network
#' # Generate the matrix of connection probabilities
#' cmat = matrix(
#'   c(
#'     30, 0.05, 0.05,
#'     0.05, 30, 0.05,
#'     0.05, 0.05, 30
#'   ),
#'   ncol = 3,
#'   byrow = TRUE
#' )
#' pmat = cmat / n
#'
#' # Creating the n x n adjacency matrix
#' adj <- matrix(0, n, n)
#' for (i in 2:n) {
#'   for (j in 1:(i - 1)) {
#'     p = pmat[class[i], class[j]] # We find the probability of connection with the weights
#'     adj[i, j] = rbinom(1, 1, p) # We include the edge with probability p
#'   }
#' }
#'
#' adjsymm = adj + t(adj)
#'
#' # When class assignment is known
#' out = goftest(adjsymm, C = class, numGraphs = 100)
#'
#' chi_sq_seq = out$statistic
#' pvalue = out$p.value
#' print(pvalue)
#'
#' # Plotting histogram of the sequence of the test statistics
#' hist(chi_sq_seq, 20, xlab = "chi-square test statistics", main = NULL)
#' abline(v = chi_sq_seq[1], col = "red", lwd = 5) # adding test statistic on the observed network
#' legend("topleft", legend = paste("observed GoF = ", chi_sq_seq[1]))
#'
#' # Example 2
#'
#'#' RNGkind(sample.kind = "Rounding")
#' set.seed(1729)
#'
#' # We model a network with 3 even classes
#' n1 = 30
#' n2 = 20
#' n3 = 50
#'
#' # Generating block assignments for each of the nodes
#' n = n1 + n2 + n3
#' class = rep(c(1, 2, 3), c(n1, n2, n3))
#'
#' # Generating the adjacency matrix of the network
#' # Generate the matrix of connection probabilities
#' cmat = matrix(
#'   c(
#'     30, 0.05, 0.05,
#'     0.05, 30, 0.05,
#'     0.05, 0.05, 30
#'   ),
#'   ncol = 3,
#'   byrow = TRUE
#' )
#' pmat = cmat / n
#'
#' # Creating the n x n adjacency matrix
#' adj <- matrix(0, n, n)
#' for (i in 2:n) {
#'   for (j in 1:(i - 1)) {
#'     p = pmat[class[i], class[j]] # We find the probability of connection with the weights
#'     adj[i, j] = rbinom(1, 1, p) # We include the edge with probability p
#'   }
#' }
#'
#' adjsymm = adj + t(adj)
#'
#' # When class assignment is known
#' out = goftest(adjsymm, C = class, numGraphs = 100)
#'
#' chi_sq_seq = out$statistic
#' pvalue = out$p.value
#' print(pvalue)
#'
#' # Plotting histogram of the sequence of the test statistics
#' hist(chi_sq_seq, 20, xlab = "chi-square test statistics", main = NULL)
#' abline(v = chi_sq_seq[1], col = "red", lwd = 5) # adding test statistic on the observed network
#' legend("topleft", legend = paste("observed GoF = ", chi_sq_seq[1]))
#'
#' # Application on real dataset: Testing on the Zachary's Karate Club Data
#'
#' set.seed(100000)
#'
#' data("zachary")
#'
#' d = zachary # the Zachary's Karate Club data set
#'
#' # the adjacency matrix
#' A_zachary = as.matrix(d[1:34, ])
#' colnames(A_zachary) = 1:34
#'
#' # obtaining the graph from the adjacency matrix above
#' g_zachary = igraph::graph_from_adjacency_matrix(A_zachary, mode = "undirected", weighted = NULL)
#'
#' # plotting the graph (network) obtained
#' plot(g_zachary,
#' main = "Network (Graph) for the Zachary's Karate Club data set; reference clustering")
#'
#' # block assignments
#' K = 2 # no. of blocks
#'
#' n1 = 10
#' n2 = 24
#' n = n1 + n2
#'
#' # known class assignments
#' class = rep(c(1, 2), c(n1, n2))

#' # goodness-of-fit tests for the Zachary's Karate Club data set
#' out_zachary = goftest(A_zachary, C = class, numGraphs = 100)
#'
#' chi_sq_seq = out_zachary$statistic
#' pvalue = out_zachary$p.value
#' print(pvalue)
#'
#' # Plotting histogram of the sequence of the test statistics
#' hist(chi_sq_seq, 20, xlab = "chi-square test statistics", main = NULL)
#' abline(v = chi_sq_seq[1], col = "red", lwd = 5) # adding test statistic on the observed network
#' legend("topleft", legend = paste("observed GoF = ", chi_sq_seq[1]))
#'
#' @references
#' Karwa et al. (2023). "Monte Carlo goodness-of-fit tests for degree corrected and related stochastic blockmodels",
#' \emph{Journal of the Royal Statistical Society Series B: Statistical Methodology},
#' <https://doi.org/10.1093/jrsssb/qkad084>
#'
#' @references
#' Qin, T., and Rohe, K. (2013). "Regularized spectral clustering under the degree-corrected stochastic blockmodel",
#' \emph{Advances in neural information processing systems},
#' <https://proceedings.neurips.cc/paper_files/paper/2013/file/0ed9422357395a0d4879191c66f4faa2-Paper.pdf>
#'
#' @references
#' Lei, J., & Rinaldo, A. (2015). "Consistency of spectral clustering in stochastic block models",
#' \emph{The Annals of Statistics},
#' <https://doi.org/10.1214/14-AOS1274>
#'
#' @references
#' Li T, Levina E, & Zhu J (2021). "randnet: Random Network Model Estimation, Selection and Parameter Tuning",
#' \emph{R packageversion 0.3},
#' <https://CRAN.R-project.org/package=randnet>
#'
#' @references
#' Ghosh, S (2022). “MCGoSBM: Monte Carlo Goodness of Fit Tests for Stochastic Block Models”,
#' \emph{Github Repository},
#' <https://github.com/GhoshSoham/MCGoFSBM/tree/main>

goftest = function(A, K = NULL, C = NULL, numGraphs = 100) {
  # The true model is beta SBM
  # Testing whether the observed graph follows beta SBM using chi square goodness of fit test for network data

  # Some compatibility checks and error message
  # Check whether the input A is a matrix

  if (!is.matrix(A) | !is.numeric(A)) {

    stop("A should be an adjacency matrix with numeric entry of the graph.")

  }

  # Check whether the graph corresponding to A is an undirected
  if (!isSymmetric.matrix(A)) {

    stop("A should be a square symmetric matrix.")

  }

  # Check whether the graph corresponding to A is an unweighted
  if (!all(A %in% c(0, 1))) {

    stop("A can only contain 0's and 1's.")

  }

  # Check whether the graph corresponding to A has no self loops
  if (!all(diag(A) == 0)) {

    stop("All the diagonal entries of A should be 0.")

  }

  # Check whether numGraphs is numeric and positive
  if (!is.numeric(numGraphs) || numGraphs <= 0 || numGraphs %% 1 != 0) {

    stop("numGraphs, number of graphs to sample should be a positive natural number.")

  }

  # Check whether C or K is provided
  if (is.null(C) & is.null(K)) {

    stop("Either block assignment or no of blocks should be provided.")

  }

  # If K is not provided, getting no of blocks from C
  if (is.null(K)) {

    # Getting no of blocks from the block assignment vector
    K <- length(unique(C))

  }

  # Check whether K is numeric and positive integer
  if (!is.numeric(K) | K %% 1 != 0) {

    stop("No of blocks K should be a positive natural number.")

  }

  # No of blocks K should be at least 2
  if (K < 2) {

    stop("No of blocks K should be at least 2.")

  }

  # If C is not provided, estimate C using value of K, the no of blocks
  if (is.null(C)) {

    # Getting block assignment for each node from adjacency matrix when block assignment is not provided
    C = block_est(A, K)

  }

  # Check whether C is numeric
  if (!is.numeric(C)) {

    stop("All the elements of C should be numeric.")

  }

  # Check whether all the elements of block assignment are from 1:K
  if (!all(C %in% 1:K)) {

    stop("C can only contain values from 1 to K (no of blocks).")

  }

  # Check whether there is at least one node from each of the block
  if (length(unique(C)) != K) {

    stop("All the blocks should have atleast one node.")

  }

  # Getting dimension information
  n = nrow(A)

  # Check whether length of C is compatible with dimension of A
  if (length(C) != n) {

    stop("The C should have same length as no of rows in A.")

  }

  # Getting an igraph (an undirected and unweighted) object from the input adjacency matrix
  G_obs = igraph::graph_from_adjacency_matrix(A, mode = "undirected", weighted = NULL)

  # Calculate estimate of the parameter from observed graph
  # which will remain same after generating a new graph on same fiber
  p_mle = get_mle(G_obs, C)

  # It will store GoF test statistic on graphs
  chi_seqR = rep(0, numGraphs)
  G = G_obs

  # Storing the first entry of chi_seq as test-stat on observed graph
  chi_seqR[1] = round(graphchi(G, C, p_mle), 2)
  for (i in 2:numGraphs) {

    # Sampling a new graph
    G_current = sample_a_move(C, G)

    # Computing GoF test statistic on new sampled graph
    chi_seqR[i] = round(graphchi(G_current, C, p_mle), 2)
    G = G_current

  }

  # pvalue i.e, proportion of sampled grpahs has larger GoF statistic than observed one
  pvalue = mean(chi_seqR > chi_seqR[1])

  # Output:
  # chi_seq: sequence of chi square test statistics on the sampled graphs
  # pvalue: estimated p-value when true model is ERSBM
  return(list(statistic = chi_seqR, p.value = pvalue))
}
